File: src/usecases/setPackingInfo.js
/* exported setPackingInfoFromActiveRow_, submitPackingInfoFromDialog_ */

function _parseCartonInput(inputText) {
  const lines = String(inputText || '').split('\n').filter(line => line.trim());
  const cartons = [];

  for (const line of lines) {
    const match = line.match(/^(\d+(?:-\d+)?)\s*[：:]\s*(\d+(?:\.\d+)?)\s*[*×x]\s*(\d+(?:\.\d+)?)\s*[*×x]\s*(\d+(?:\.\d+)?)\s+(\d+(?:\.\d+)?)\s*(?:KG|kg)?$/i);
    if (!match) {
      throw new Error(`解析できない行: "${line}"\n形式: 箱番号：長さ*幅*高さ 重さKG\n例: 1-2：60*40*32 29.1KG`);
    }

    const [, boxRange, lengthCm, widthCm, heightCm, weightKg] = match;
    const dimensions = {
      length: parseFloat(lengthCm),
      width: parseFloat(widthCm),
      height: parseFloat(heightCm)
    };
    const weight = parseFloat(weightKg);

    if (boxRange.includes('-')) {
      const [start, end] = boxRange.split('-').map(Number);
      for (let i = start; i <= end; i++) {
        cartons.push({ boxNumber: i, dimensions, weight });
      }
    } else {
      cartons.push({ boxNumber: parseInt(boxRange, 10), dimensions, weight });
    }
  }

  cartons.sort((a, b) => a.boxNumber - b.boxNumber);
  return cartons;
}

function _cmToInches(cm) {
  return Math.round(cm / 2.54 * 100) / 100;
}

function _kgToLbs(kg) {
  return Math.round(kg * 2.20462 * 100) / 100;
}

function _extractInboundPlanIdFromCell(cellValue) {
  const value = String(cellValue || '');
  const wfMatch = value.match(/wf[a-f0-9-]+/i);
  if (wfMatch) {
    return wfMatch[0];
  }
  if (/^wf[a-f0-9-]+$/i.test(value.trim())) {
    return value.trim();
  }
  throw new Error(`納品プランIDを特定できません: "${value}"`);
}

function setPackingInfoFromActiveRow_() {
  const config = getEnvConfig();
  const sheet = new PurchaseSheet(config.PURCHASE_SHEET_NAME);
  sheet.getActiveRowData();

  if (sheet.data.length === 0) {
    throw new Error('選択された行がありません');
  }

  const row = sheet.data[0];
  const planCellValue = row.get('納品プラン');
  const inboundPlanId = _extractInboundPlanIdFromCell(planCellValue);

  console.log(`[setPackingInfo] inboundPlanId=${inboundPlanId}`);

  _showPackingInfoDialog_(inboundPlanId);
}

function _showPackingInfoDialog_(inboundPlanId) {
  const html = HtmlService.createHtmlOutput(`
    <div style="font-family: Arial, sans-serif; padding: 8px;">
      <h3>荷物情報を入力</h3>
      <div style="margin:8px 0;">InboundPlanId: <code>${inboundPlanId}</code></div>
      <div style="margin:8px 0;">
        <label>
          荷物情報（1行に1パターン）：
          <br>
          <textarea id="cartonInput" rows="10" cols="50" placeholder="1-2：60*40*32　29.1KG&#10;3：30*40*50　25.8KG&#10;4：30*40*50　28.2KG"></textarea>
        </label>
      </div>
      <div style="margin:8px 0; font-size:12px; color:#666;">
        形式: 箱番号：長さ*幅*高さ 重さKG<br>
        箱番号は「1-2」のように範囲指定も可能（同じサイズ・重さの場合）
      </div>
      <div style="margin-top:12px;">
        <button onclick="submitCartonInfo()">送信</button>
        <button onclick="google.script.host.close()">キャンセル</button>
      </div>
      <pre id="status" style="margin-top:12px; background:#f6f6f6; padding:8px;"></pre>
    </div>

    <script>
      const inboundPlanId = ${JSON.stringify(inboundPlanId)};

      function submitCartonInfo() {
        const input = document.getElementById('cartonInput').value;
        if (!input.trim()) {
          document.getElementById('status').textContent = '荷物情報を入力してください';
          return;
        }

        document.getElementById('status').textContent = '送信中...';
        google.script.run
          .withSuccessHandler((res) => {
            document.getElementById('status').textContent =
              '完了\\n' + JSON.stringify(res, null, 2);
          })
          .withFailureHandler((err) => {
            document.getElementById('status').textContent =
              'エラー\\n' + (err && err.message ? err.message : String(err));
          })
          .submitPackingInfoFromDialog(inboundPlanId, input);
      }
    </script>
  `).setWidth(600).setHeight(500);

  SpreadsheetApp.getUi().showModalDialog(html, '荷物情報入力');
}

function submitPackingInfoFromDialog_(inboundPlanId, cartonInputText) {
  const accessToken = getAuthToken();
  const creator = new InboundPlanCreator(accessToken);

  const cartons = _parseCartonInput(cartonInputText);
  console.log(`[setPackingInfo] parsed cartons: ${JSON.stringify(cartons)}`);

  const packingGroupId = creator.getPackingGroupId(inboundPlanId);
  console.log(`[setPackingInfo] packingGroupId=${packingGroupId}`);

  const items = creator.getPackingGroupItems(inboundPlanId, packingGroupId);
  console.log(`[setPackingInfo] items count=${items.length}`);

  const result = creator.setPackingInformation(inboundPlanId, packingGroupId, cartons, items);
  console.log(`[setPackingInfo] result: ${JSON.stringify(result)}`);

  return result;
}

File: python/src/usecases/set_packing_info.py
from __future__ import annotations

import logging
import re
from typing import Any

import click

from shared.config import AppConfig
from infrastructure.amazon.auth import get_auth_token
from infrastructure.amazon.inbound_plan_creator import InboundPlanCreator
from infrastructure.spreadsheet.base_sheets_repository import BaseSheetsRepository
from infrastructure.spreadsheet.purchase_sheet import PurchaseSheet

logger = logging.getLogger(__name__)


def set_packing_info(config: AppConfig, repo: BaseSheetsRepository, row_numbers: list[int]) -> None:
    access_token = get_auth_token(config.api_key, config.api_secret, config.refresh_token)
    sheet = PurchaseSheet(repo, config.sheet_id, config.purchase_sheet_name)
    sheet.get_rows_by_numbers(row_numbers)
    plan_cell = str(sheet.data[0].get("納品プラン") or "").strip() if sheet.data else ""
    inbound_plan_id = extract_inbound_plan_id(plan_cell)
    if not inbound_plan_id:
        raise RuntimeError("納品プランIDが取得できません")
    click.echo(f"納品プランID: {inbound_plan_id}")
    click.echo("箱情報を入力してください（例: 1-2：60*40*32 29.1KG）")
    click.echo("空行で入力終了:")
    lines: list[str] = []
    while True:
        line = input()
        if not line.strip():
            break
        lines.append(line)
    carton_text = "\n".join(lines)
    cartons = parse_carton_input(carton_text)
    if not cartons:
        raise RuntimeError("箱情報がパースできません")
    creator = InboundPlanCreator(access_token)
    packing_group_id = creator.get_packing_group_id(inbound_plan_id)
    items = creator.get_packing_group_items(inbound_plan_id, packing_group_id)
    body = build_packing_body(packing_group_id, cartons, items)
    creator.set_packing_information(inbound_plan_id, body)
    click.echo("梱包情報の送信が完了しました")


PLAN_ID_PATTERN = r"wf[0-9a-fA-F]{8}(?:-[0-9a-fA-F]{4}){3}-[0-9a-fA-F]{12}|wf[a-zA-Z0-9]+"
LABELING_PREP_TYPE = "ITEM_LABELING"


def extract_inbound_plan_id(cell_value: str) -> str:
    match = re.search(rf"wf=({PLAN_ID_PATTERN})", cell_value)
    if match:
        return match.group(1)
    match = re.search(rf"({PLAN_ID_PATTERN})", cell_value)
    return match.group(1) if match else ""


def parse_carton_input(text: str) -> list[dict[str, Any]]:
    cartons: list[dict[str, Any]] = []
    for line in text.strip().split("\n"):
        line = line.strip()
        if not line:
            continue
        match = re.match(
            r"(\d+(?:-\d+)?)\s*[：:]\s*(\d+(?:\.\d+)?)\s*[*×x]\s*(\d+(?:\.\d+)?)\s*[*×x]\s*(\d+(?:\.\d+)?)\s+(\d+(?:\.\d+)?)\s*[Kk][Gg]",
            line,
        )
        if not match:
            logger.warning("パース失敗: %s", line)
            continue
        box_range = match.group(1)
        length_cm = float(match.group(2))
        width_cm = float(match.group(3))
        height_cm = float(match.group(4))
        weight_kg = float(match.group(5))
        if "-" in box_range:
            start, end = box_range.split("-")
            count = int(end) - int(start) + 1
        else:
            count = 1
        cartons.append({
            "count": count,
            "length": length_cm,
            "width": width_cm,
            "height": height_cm,
            "weight": weight_kg,
        })
    return cartons


def build_packing_body(
    packing_group_id: str, cartons: list[dict[str, Any]], items: list[dict[str, Any]],
) -> dict[str, Any]:
    box_items = [
        {
            "msku": item.get("msku", ""),
            "quantity": item.get("quantity", 0),
            "labelOwner": item.get("labelOwner", "SELLER"),
            "prepOwner": _prep_owner_of(item),
        }
        for item in items
    ]
    package_groups = [
        {
            "packingGroupId": packing_group_id,
            "boxes": [{
                "contentInformationSource": "BOX_CONTENT_PROVIDED",
                "weight": {"unit": "KG", "value": carton["weight"]},
                "dimensions": {
                    "unitOfMeasurement": "CM",
                    "length": carton["length"],
                    "width": carton["width"],
                    "height": carton["height"],
                },
                "quantity": carton["count"],
                "items": box_items,
            }],
        }
        for carton in cartons
    ]
    return {"packageGroupings": package_groups}


def _prep_owner_of(item: dict[str, Any]) -> str:
    for instruction in item.get("prepInstructions") or []:
        if instruction.get("prepType") == LABELING_PREP_TYPE:
            continue
        if instruction.get("prepOwner"):
            return str(instruction["prepOwner"])
    return "NONE"
